### DISCUSSION TITLE
Plan for animation component for GameObjects (DOTweenAnimation equivalent)?

### DISCUSSION THREAD

**Discussion by laicasaane** [KyryloKuzyk/PrimeTween#36] [Ideas]
Currently DOTween Pro provides a DOTweenAnimation component so we can configure tweening with no code on GameObjects. Will you add something like this to Prime Tween in the near future? Because of this I've not switched to Prime Tween for the commercial project I'm currently working on.

**Comment by KyryloKuzyk** [comment]
I do have a plan to release a paid PrimeTween PRO with an equivalent to the DOTweenAnimation. However, DOTweenAnimation is quite limited (it can play only one animation), and I don't want to rush with the same limited feature in PrimeTween. As I see it, PrimeTween PRO users should be able to create complex sequences from the inspector. Creating sequences from the inspector is not a trivial task from the UX/UI standpoint, and I'm still evaluating how to do it right. I don't want to rush with a half-baked solution, so I can't give any time estimates for now. 

As an alternative to the DOTweenAnimation component, PrimeTween offers the [Inspector Integration](https://github.com/KyryloKuzyk/PrimeTween#inspector-integration) feature. I would say Inspector integration already covers the majority of DOTweenAnimation usages: it lets you tweak all animation properties from the Inspector without the need to recompile your project. And for more complex use cases, I believe it's still better (and more reliable) to use C#.

**Reply by laicasaane** [reply]
In my project I only need simple functionality of DOTweenAnimation. I see it as a reusable component for trivial cases where I don't want to create components or write specialized code just for tweening GOs.

**Reply by KyryloKuzyk** [reply]
@laicasaane But you do have to declare a reference `[SerializeField] DOTweenAnimation myAnimation;`, so you can then call `myAnimation.DORestart();`, right? This is also a specialized code that you have to write for every animation (unless this animation is played automatically on scene start or is called from UnityEvent in a scene, which are quite rare use cases in my opinion). 

In this case, declaring `[SerializeField] TweenSettings<float> moveSettings;`, then calling `Tween.PositionY(transform, moveSettings);` would result in the same amount of specialized code.

**Reply by laicasaane** [reply]
The restart method is called by a manager actually. So I don't have to write any code just for this.
![image](https://github.com/KyryloKuzyk/PrimeTween/assets/1594982/19a2fffc-b8e8-40a7-a8b6-f0212b2c94a8)

**Reply by KyryloKuzyk** [reply]
@laicasaane Thank you for providing more context! Indeed, PrimeTween doesn't currently have a built-in alternative for this use case. 

This use case can be easily addressed with a custom component. Yes, you'll have to create a component yourself, but you'll be in full control over how it works. Here is an example of scale/translate/rotate animations. This script can be easily extended to support all other animation types that your project needs:
```csharp
using PrimeTween;
using UnityEngine;

public class AnimationComponent : MonoBehaviour {
    [SerializeField] AnimationType animationType;
    [SerializeField] bool playOnEnable = true;
    [SerializeField] TweenSettings<Vector3> tweenSettings;

    void OnEnable() {
        if (playOnEnable) {
            Play();
        }
    }

    public Tween Play() {
        switch (animationType) {
            case AnimationType.Scale:
                return Tween.Scale(transform, tweenSettings);
            case AnimationType.Position:
                return Tween.Position(transform, tweenSettings);
            case AnimationType.LocalPosition:
                return Tween.LocalPosition(transform, tweenSettings);
            case AnimationType.Rotation:
                return Tween.Rotation(transform, tweenSettings);
            case AnimationType.LocalRotation:
            default:
                return Tween.LocalRotation(transform, tweenSettings);
        }
    }

    enum AnimationType {
        Scale,
        Position, LocalPosition,
        Rotation, LocalRotation
    }
}
```
